src/hotspot/share/opto/vectornode.hpp

```diff
@@ -855,20 +855,44 @@ class LoadVectorNode : public LoadNode {
   uint element_size(void) { return type2aelembytes(vect_type()->element_basic_type()); }
 };
 
-//------------------------------LoadVectorGatherNode------------------------------
+//------------------------------GatherNode------------------------------
 // Load Vector from memory via index map
-class LoadVectorGatherNode : public LoadVectorNode {
+class GatherNode : public LoadVectorNode {
  public:
-  LoadVectorGatherNode(Node* c, Node* mem, Node* adr, const TypePtr* at, const TypeVect* vt, Node* indices)
+  GatherNode(Node* c, Node* mem, Node* adr, const TypePtr* at, const TypeVect* vt, Node* indices, Node* scale)
     : LoadVectorNode(c, mem, adr, at, vt) {
-    init_class_id(Class_LoadVectorGather);
-    assert(indices->bottom_type()->is_vect(), "indices must be in vector");
+    init_class_id(Class_Gather);
+    const TypeVect* indices_vt = indices->bottom_type()->is_vect();
+    assert(indices_vt->length() == vt->length(), "indices must have the same length");
     add_req(indices);
-    assert(req() == MemNode::ValueIn + 1, "match_edge expects that last input is in MemNode::ValueIn");
+    add_req(scale);
+    assert(req() == MemNode::ValueIn + 2, "match_edge expects that last input is in MemNode::ValueIn");
   }
 
-  virtual int Opcode() const;
   virtual uint match_edge(uint idx) const { return idx == MemNode::Address || idx == MemNode::ValueIn; }
+
+  static int opcode(BasicType bt);
+  static GatherNode* make(int opc, Node* c, Node* mem, Node* adr, const TypePtr* at, const TypeVect* vt, Node* indices, Node* scale);
+};
+
+class GatherINode : public GatherNode {
+ public:
+  GatherINode(Node* c, Node* mem, Node* adr, const TypePtr* at, const TypeVect* vt, Node* indices, Node* scale)
+    : GatherNode(c, mem, adr, at, vt, indices, scale) {
+    assert(indices->bottom_type()->is_vect()->element_basic_type() == T_INT, "");
+  }
+
+  virtual int Opcode();
+};
+
+class GatherLNode : public GatherNode {
+ public:
+  GatherLNode(Node* c, Node* mem, Node* adr, const TypePtr* at, const TypeVect* vt, Node* indices, Node* scale)
+    : GatherNode(c, mem, adr, at, vt, indices, scale) {
+    assert(indices->bottom_type()->is_vect()->element_basic_type() == T_LONG, "");
+  }
```

Comment:
We can also get the index type info from the index vector. So is it possible that we use a single node for the gather op, and do the different codegen in ad files based on the index element type? The same to the scatter op.

Reply:
I'm not sure that all architectures will support 32-bit indices and 64-bit indices similarly. So I think split them make it easier querying for backend support. In x86 a 64-bit index gather support depends on the vector length while 32-bit one depends on the vector length in byte.

Reply:
OK. Make sense to me. And an alternative way is to adding another matcher method to finish the check. But the change sames also complex. So maybe we could rename the node to `LoadGatherIndexINode/LoadGatherIndexLNode` ?